init.c

```diff
@@ -184,9 +184,31 @@ void waitMultiDeviceEvents(lua_State *L, THCState *state, int arg,
   lua_pop(L, 1);
 }
 
+static int optSetDevice(lua_State *L, int param_idx, const char *func_name)
+{
+  THCState *state = cutorch_getstate(L);
+  int oldDevice;
+  THCudaCheck(cudaGetDevice(&oldDevice));
+  int device = luaL_optint(L, param_idx, -10);
+  if (device != -10) {
+    THCState_setDevice(state, device - 1); /* zero indexed */
+  } else if (THCState_getDeviceMode(state) != THCStateDeviceModeManual) {
+      THError("%s: Current deviceID=0 and no device provided. Provide the deviceID as argument %d",
+          func_name, param_idx);
+  }
+  return oldDevice;
+}
+
+/*
+   Usage:
+   cutorch.synchronize([device])
+   Synchronizes the current or specified device with the host.
+*/
 static int cutorch_synchronize(lua_State *L)
 {
+  int oldDevice = optSetDevice(L, 1, "synchronize");
   THCudaCheck(cudaDeviceSynchronize());
+  THCState_setDevice(cutorch_getstate(L), oldDevice);
```

Comment:
I feel it would be less error-prone. You can't accidentally forget to reset the device. If you call it something like `withOptDevice` it should be fairly clear what's going on.